### ISSUE TITLE
RFC: CIFlow for PyTorch: a flexible CI dispatcher that enables better PR experience with configurable testing validation set

### ISSUE THREAD

**Issue by zhouzhuojie** [pytorch/pytorch#61888]
# 1. Introduction
We are going to introduce a PyTorch OSS CI Dispatcher - CIFlow, which is going to dispatch GitHub Actions CI workflows based on various PR contexts and user instructions. It's like the target determination but for CI workflows to help us dynamically optimize and determine what workflows to run at PR time. There are two components of the implementation:

- A github event-driven dispatcher called "ciflow" (deployed as a “broker action” on probot).
- A modular DSL (reusing existing jinja2) for the workflows and jobs' common configuration.

The goal of the CIFlow is to increase the modularity, flexibility, and efficiency of the PyTorch’s CI system. Recently, the pain points of OSS CI are identified as one of the urgent efforts that can improve developers’ productivity.

Note, for simplicity, we call it `GHA` for GitHub Actions. And this RFC is primarily targeting the workflows on GHA, though if we're using probot as the root dispatcher, we can easily extend the idea to circleci workflows.

# 2. Problem Space
Technical Challenges of the existing [probot + circleci + gha + jenkins] system:

1. _**Hard to introduce a flexible dispatcher within the current structure of CI.**_ We may need the flexibility to control what validation set we want to run at the PR time. Currently we have about 100 CI checks (workflows and jobs across GHA and circleci) at PR time. For example, people might only want to validate lint if it's just a specific lint configuration change; or people might want to validate extra windows tests in addition to the default set that runs on every PR; or a smart enough CI system that can figure out how to test no more no less with high signals-to-noise ratio. Due to the tight coupling of GHA and its PR system, currently it's challenging to introduce dispatch algorithms that can dynamically assign workflows, it's even more challenging to move towards a CI vendor dynamic switch framework. 
2. _**Hard to reason what workflows are running under what conditions.**_ What workflows are triggered is hard to reason about. Ideally, we'd like to see the "test plan" showing upfront to the users with a clear understanding of the validation set.
The entry points of tests are hard to find. Tests are wrapped with multiple layers of scripts spreading across many `yaml`, `.sh` and `.py` files, even multiple repos. We may want to refactor the entry points of the tests, and then modularize the CI configuration to enable a flexible dispatch CI system. We’d like to see a modular DSL of CI component definition, CI entry points, and their event-driven dependency graph. We have many condition checks and environment variables that are gating which platform/arch/configuration set we should use. However, it is difficult to know why and how they are configured. Sometimes, CI vendors have distinct logic to express triggers and conditions. 

# 3. Implementation Proposal

There will be two components to achieve the flexibility we want:
- A root dispatcher that can compile the "test plan" - i.e. based on the PR context and user instruction, craft the "test plan" that the workflows can react on.
- A trigger action **that's rarely used** by normal PR users, bots, or reviewers (e.g. `on.pull_request [unassigned]` or `on.pull_request [unlabeled]`), and the GHA workflows understand the event context and knows how to test its jobs accordingly. The reason we can't bundle this into `[labeled]` event or other API driven design is that the event that triggers the PR has to be one of the limited subset of the actions that GHA supports.
  - See GHA doc about `on.pull_request` triggers: https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request
  - See an open feature request about doing label filters: https://github.com/github/feedback/discussions/4679 

A demo of the architecture (yml names may not match what we have):
![image](https://user-images.githubusercontent.com/658840/126261336-630bf00f-ce42-4e3e-8e64-92c72ed8837e.png)

# 4. User Experience
A dedicated test plan comment block will be created automatically by `pytorchbot`(probot), and PR authors and reviewers can leverage that as the source of truth of what kind of workflows will be run for the PR. 

An early demo of the UI of the comment block (will be changed):
![image](https://user-images.githubusercontent.com/658840/126262023-f7d795e7-7022-4e4f-8b14-89919f6492d5.png)

### 4.1 Craft test plan
The test plan will be configured by two components
- The set of the labels that maps to workflow yml files (with sane defaults, and user input)
- The workflows can react to events (e.g. `on.pull_request [unassigned]` or `on.pull_request [unlabeled]`) with the labels in the event context 

Initial labels design (i.e. the labels to workflows mapping):

```
label [ciflow/default]:
  - pytorch-linux-bionic-py3.8-gcc9-coverage.yml
  - pytorch-linux-xenial-py3.6-gcc5.4.yml
  - pytorch-win-vs2019-cpu-py3.yml
  - pytorch-win-vs2019-cuda10-cudnn7-py3.yml
  - ... (add more)
label [ciflow/master]:
  - pytorch-libtorch-linux-xenial-cuda10.2-cudnn7-py3.6-gcc7.yml
  - pytorch-libtorch-linux-xenial-cuda11.1-cudnn8-py3.6-gcc7.yml
  - pytorch-linux-bionic-cuda10.2-cudnn7-py3.9-gcc7.yml
  - pytorch-linux-xenial-cuda10.2-cudnn7-py3.6-gcc7.yml
  - pytorch-linux-xenial-cuda11.1-cudnn8-py3.6-gcc7.yml
  - pytorch-linux-xenial-py3.6-gcc7-bazel-test
  - ... (add more)
label [ciflow/slow]:
  - pytorch-linux-xenial-cuda10.2-cudnn7-py3.6-gcc7.yml
  - ... (add more)
```

### 4.2 Trigger the workflows
After that, users can either 1) add or remove labels themselves, or 2) change the comment block so that probot can add the labels automatically. Note we may want to carefully consider the github api for label editing or comments editing permission here. And then, probot will trigger the event so that all the GHA workflows will be running based on the event context.

# 5. Rollout Plan

 1. Manual Phase
step 1. Add 'ciflow/default' label to the PR
step 2. Once there's an [unassigned] event from PR, it should rerun
step 3. Remove 'ciflow/default' label
step 4. Trigger the [unassigned] event again, it should not rerun
2. Probot Phase 1 (manual on 1 workflow)
step 1. Probot automatically add labels based on the context
step 2. Manually let probot trigger [unassigned] event
3. Probot Phase 2 (auto on 1 workflows)
step 1. Modify the workflows so that they only listen on [unassigned] events
step 2. Probot automatically adds labels automatically based on the context
step 3. Probot automatically triggers [unassigned] event
4. Probot Phase 3 (auto on many workflows)
step 1. Enable it for all workflows

PRs:
- [x] pytorch/pytorch changes
  - [x] https://github.com/pytorch/pytorch/pull/61879
  - [x] https://github.com/pytorch/pytorch/pull/61886
  - [x] slow rollout to more workflows
- [x] pytroch/pytorch-probot changes
  - [x] refactor and testing setup https://github.com/pytorch/pytorch-probot/pull/40
  - [x] context collection https://github.com/pytorch/pytorch-probot/pull/40
  - [x] comment block generation https://github.com/pytorch/pytorch-probot/pull/51
  - [x] add labels https://github.com/pytorch/pytorch-probot/pull/40
  - [x] trigger [unassigned] events https://github.com/pytorch/pytorch-probot/pull/40
- [ ] roll out to more users

---- 
FB internal RFC link: https://docs.google.com/document/d/1uAbskULQvy8LnKrQ5gf86GzDyn42qwOMEqiEUUp8Xt0

cc @ezyang @seemethere @malfet @walterddr @pytorch/pytorch-dev-infra

**Comment by suo** [comment]
This is cool! Some random thoughts/questions:
- How will developers change or add to the ruleset for determining what CI jobs run? Will it be some file in-tree?
- It would be super awesome if we could get a tool that explained *why* a given CI job ran for a PR.

**Comment by zhouzhuojie** [comment]
- Yes, ruleset will be checked in, could be in any declarative format, e.g. either a YAML or simple Python dict to indicate what workflows can react to PR context. We will start with simple things like `labels` to `workflows` mapping, and then iterate on more. 
- Explicit "test plan" will be updated by the probot

**Comment by zhouzhuojie** [comment]
This was already live in pytorch! Closing this RFC
